File: src/II062.cpp
#include <array>
#include <cstddef>
#include <iostream>
#include <vector>
using namespace std;

class Trie {
  array<Trie*, 26> children;
  bool is_end;

  Trie* searchPrefix(const string& prefix) {
    Trie* p = this;
    for (char ch : prefix) {
      ch -= 'a';
      if (p->children[ch] == nullptr) return nullptr;
      p = p->children[ch];
    }
    return p;
  }

public:
  /** Initialize your data structure here. */
  Trie(bool is_end = false) : children(), is_end(is_end) {} // array必须显式初始化

  /** Inserts a word into the trie. */
  void insert(string word) {
    Trie* p = this;
    for (char ch : word) {
      ch -= 'a';
      if (p->children[ch] == nullptr) {
        p->children[ch] = new Trie();
      }
      p = p->children[ch];
    }
    p->is_end = true;
  }

  /** Returns if the word is in the trie. */
  bool search(string word) {
    auto res = searchPrefix(word);
    return res != nullptr && res->is_end;
  }

  /** Returns if there is any word in the trie that starts with the given prefix. */
  bool startsWith(string prefix) { return searchPrefix(prefix) != nullptr; }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

int main() {
  Trie* obj = new Trie();
  obj->insert("apple");
  bool param_2 = obj->search("app");
  bool param_3 = obj->startsWith("app");
  cout << param_2 << ' ' << param_3 << endl;
}

File: src/II062.rs
#[derive(Default)]
struct Trie {
    children: [Option<Box<Trie>>; 26],
    is_end: bool,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Trie {
    /** Initialize your data structure here. */
    fn new() -> Self {
        Default::default()
    }

    /** Inserts a word into the trie. */
    fn insert(&mut self, word: String) {
        let mut p = self;
        for i in word.bytes().map(|ch| (ch - b'a') as usize) {
            // p = p.children[i].get_or_insert(Box::new(Self::new()));
            p = p.children[i].get_or_insert_with(|| Box::new(Trie::new())); // 这比上面快很多，为什么？
        }
        p.is_end = true;
    }

    /** Returns if the word is in the trie. */
    fn search(&self, word: String) -> bool {
        let mut p = self;
        for i in word.bytes().map(|ch| (ch - b'a') as usize) {
            match &p.children[i] {
                Some(v) => p = v,
                _ => return false,
            }
        }
        p.is_end
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    fn starts_with(&self, prefix: String) -> bool {
        let mut p = self;
        for i in prefix.bytes().map(|ch| (ch - b'a') as usize) {
            match &p.children[i] {
                Some(v) => p = v,
                _ => return false,
            }
        }
        true
    }
}

// /**
//  * `&self` means the method takes an immutable reference.
//  * If you need a mutable reference, change it to `&mut self` instead.
//  */
// impl Trie {
//     /** Initialize your data structure here. */
//     fn new() -> Self {
//         Default::default()
//     }

//     /** Inserts a word into the trie. */
//     fn insert(&mut self, word: String) {
//         let mut p = self;
//         for &ch in word.as_bytes() {
//             p = p.children[(ch - b'a') as usize].get_or_insert(Box::new(Self::new()));
//         }
//         p.is_end = true;
//     }

//     /** Returns if the word is in the trie. */
//     fn search(&self, word: String) -> bool {
//         let mut p = self;
//         for &ch in word.as_bytes() {
//             if let Some(ref node) = p.children[(ch - b'a') as usize] {
//                 p = node;
//             } else {
//                 return false;
//             }
//         }
//         p.is_end
//     }

//     /** Returns if there is any word in the trie that starts with the given prefix. */
//     fn starts_with(&self, prefix: String) -> bool {
//         let mut p = self;
//         for &ch in prefix.as_bytes() {
//             if let Some(ref node) = p.children[(ch - b'a') as usize] {
//                 p = node;
//             } else {
//                 return false;
//             }
//         }
//         true
//     }
// }
